src/mito2/src/memtable/merge_tree/merger.rs

```diff
@@ -0,0 +1,633 @@
+                                to_yield
+                            } else {
+                                let to_yield = top_node.current_item().slice(0..pos);
+                                top_node.skip(pos)?;
+                                to_yield
+                            }
+                        }
+                        Err(pos) => {
+                            // no duplicated timestamp
+                            let to_yield = top_node.current_item().slice(0..pos);
+                            top_node.skip(pos)?;
+                            to_yield
+                        }
+                    };
+                    self.current_item = Some(res);
+                }
+            } else {
+                // top is the only node left.
+                self.current_item = Some(top_node.fetch_next()?);
+            }
+            if top_node.is_valid() {
+                self.heap.push(top_node);
+            }
+        } else {
+            // heap is empty
+            self.current_item = None;
+        }
+        Ok(())
+    }
+
+    /// Returns current item held by merger.
+    pub(crate) fn current_item(&self) -> &I {
+        self.current_item.as_ref().unwrap()
+    }
+}
+
+#[derive(Debug)]
+pub struct DataBatchKey {
+    pk_index: PkIndex,
+    timestamp: i64,
+}
+
+impl Eq for DataBatchKey {}
+
+impl PartialEq<Self> for DataBatchKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.pk_index == other.pk_index && self.timestamp == other.timestamp
+    }
+}
+
+impl PartialOrd<Self> for DataBatchKey {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for DataBatchKey {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.pk_index
+            .cmp(&other.pk_index)
+            .then(self.timestamp.cmp(&other.timestamp))
+            .reverse()
+    }
+}
+
+impl DataBatch {
+    fn first_row(&self) -> (i64, u64) {
+        let range = self.range();
+        let ts_values = timestamp_array_to_i64_slice(self.rb.column(1));
+        let sequence_values = self
+            .rb
+            .column(2)
+            .as_any()
+            .downcast_ref::<UInt64Array>()
+            .unwrap()
+            .values();
+        (ts_values[range.start], sequence_values[range.start])
+    }
+
+    fn last_row(&self) -> (i64, u64) {
+        let range = self.range();
+        let ts_values = timestamp_array_to_i64_slice(self.rb.column(1));
+        let sequence_values = self
+            .rb
+            .column(2)
+            .as_any()
+            .downcast_ref::<UInt64Array>()
+            .unwrap()
+            .values();
+        (ts_values[range.end - 1], sequence_values[range.end - 1])
+    }
+}
+
+impl Item for DataBatch {
+    type Key = DataBatchKey;
+
+    fn remaining(&self) -> usize {
+        self.range().len()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.is_empty()
+    }
```

Comment:
The merger doesn't use them. We don't need to define them in the Item trait.

Reply:
> The merger doesn't use them. We don't need to define them in the Item trait.

fixed in https://github.com/GreptimeTeam/greptimedb/pull/3361/commits/dc1c86a1ff2acf1c0cea69dc3960a661de2f7a23#diff-8e513b995e680c513011a438989b7551954f873b55a1a4ba602550e21d5d8094L217-L220